WAVLTree/src/WAVLTree.java

```diff
@@ -2,172 +2,207 @@
+	 * tree is empty
+	 */
+	public String max() {
+		return "42"; // to be replaced by student code
+	}
+
+	/**
+	 * public int[] keysToArray()
+	 *
+	 * Returns a sorted array which contains all keys in the tree, or an empty array
+	 * if the tree is empty.
+	 */
+	public int[] keysToArray() {
+		int[] arr = new int[42]; // to be replaced by student code
+		return arr; // to be replaced by student code
+	}
+
+	/**
+	 * public String[] infoToArray()
+	 *
+	 * Returns an array which contains all info in the tree, sorted by their
+	 * respective keys, or an empty array if the tree is empty.
+	 */
+	public String[] infoToArray() {
+		String[] arr = new String[42]; // to be replaced by student code
+		return arr; // to be replaced by student code
+	}
+
+	/**
+	 * public int size()
+	 *
+	 * Returns the number of nodes in the tree.
+	 *
+	 */
+	public int size() {
+		return 42; // to be replaced by student code
+	}
+
+	/**
+	 * public WAVLNode getRoot()
+	 *
+	 * Returns the root WAVL node, or null if the tree is empty
+	 *
+	 */
+	public WAVLNode getRoot() {
+		return null;
+	}
+
+	/**
+	 * public int select(int i)
+	 *
+	 * Returns the value of the i'th smallest key (return -1 if tree is empty)
+	 * Example 1: select(1) returns the value of the node with minimal key Example
+	 * 2: select(size()) returns the value of the node with maximal key Example 3:
+	 * select(2) returns the value 2nd smallest minimal node, i.e the value of the
+	 * node minimal node's successor
+	 *
+	 */
+
+	public String select(int i) {
+		return null;
+	}
+
+	/**
+	 * public class WAVLNode
+	 */
+	//
+	public class WAVLNode {
+		public static final int OUTER_NODE_RANK = -1;
+		public static final String OUTER_NODE_VALUE = "";
+		public static final int OUTER_NODE_KEY = -1;
+		private int key;
+		private String value;
+		private WAVLNode parent;
+		private WAVLNode right;
+		private WAVLNode left;
+		private int rank;
+		private int size; // the size field needs to be updated after inserts or rebalances.
+
+		public WAVLNode(int key, String value, WAVLNode parent, WAVLNode right, WAVLNode left, int rank) {
+			this.key = key;
+			this.value = value;
+			this.parent = parent;
+			this.right = right;
+			this.left = left;
+			this.rank = rank;
+			this.size = this.getSubtreeSize();
+		}
+
+		public WAVLNode(int key, String value) {
+			this(key, value, null, null, null, 1);
+		}
+		
+		/**
+		 * The default constructor for the WAVLNode class.
+		 * Constructs an external node.
+		 */
+		public WAVLNode() {
+			this(OUTER_NODE_KEY, OUTER_NODE_VALUE, null, null, null, OUTER_NODE_RANK);
+		}
+		public int getKey() {
+			return key; 
+		}
+
+		public String getValue() {
+			return value;
+		}
+
+		public WAVLNode getLeft() {
+			return left;
+		}
+
+		public WAVLNode getRight() {
+			return right;
+		}
+		public WAVLNode getParent() {
+			return parent;
+		}
+
+		public boolean isInnerNode() {
+			return rank == OUTER_NODE_RANK;
```

Comment:
Should this be `!=` instead of `==` since `OUTER_NODE_RANK = -1`?
```suggestion
			return rank != OUTER_NODE_RANK;
```